enhancements/installer/aws-custom-dns.md

```diff
@@ -0,0 +1,476 @@
+solution (Route53).
+- Lay the groundwork to add this capability for other cloud platforms like
+Azure and GCP.
+- Continue using the cloud based Load Balancer service for API, Ingress and
+API-Int IPs.
+
+### Non-Goals
+
+- We will not focus on Installer changes to support the custom DNS solution on
+Azure and GCP platforms. That will come in follow-up work and will rely on the
+changes introduced within this solution.
+
+- We will not use the [Kubernetes External DNS provider](https://github.com/openshift/external-dns) and/or the [OpenShift External DNS Operator](https://github.com/openshift/external-dns-operator) to program dns entries across varied DNS providers.
+The External DNS Operator relies on being able to program the customer's
+DNS solution. The customer DNS solutions in consideration here do not allow
+OpenShift to configure them.
+
+## Proposal
+
+For the OpenShift installation to make use of a custom DNS solution that is not
+configurable by OpenShift, the customer is responsible for configuring the
+DNS entries for the API and Ingress services.
+
+When it is indicated to the OpenShift installer via an input parameter that the
+custom DNS solution is in use, OpenShift does not configure the cloud DNS
+solution [Route53 for AWS](https://aws.amazon.com/route53/). OpenShift will
+continue to configure and manage the cloud load balancing solution [Elastic
+Load Balancer](https://aws.amazon.com/elasticloadbalancing/) for all its
+services.
+
+OpenShift will start static CoreDNS pods to provide DNS resolution for API,
+Ingress API and Ingress services that are essentail for cluster creation.
+Although not identical, this approach leverages learning from OpenShift's
+approach to providing these services for on-prem platforms.
+
+### Workflow Description
+
+1. For an OpenShift administrator to use a custom DNS solution, they would
+have to enable a config option in the install configuration. This would
+cause the Installer and the Ingress Operator to skip configuring the cloud
+DNS with entries for API, Internal API and Ingress Services. 
+
+2. The Installer and the Ingress Operator will continue to configure the cloud
+Load Balancers on the public subnets of Public Hosted Zones for `api` and
+`*.apps`. They will continue to configure cloud Load Balancers on the private
+subnets of Private Hosted Zones for `api-int` and `*.apps` for in-cluster
+access to the Ingress service.
+
+3. The Installer also creates a ConfigMap(CloudLBConfig) to store LB API
+addresses to be used later for configuration of API and API-Int DNS records
+on the in-cluster and customer and customer DNS solutions. 
+
+4. The Installer instructs the machine-config-operator(MCO) to start a static
+CoreDNS pod on the bootstrap and master nodes to provide DNS resolution for
+the API and API-Int servers. MCO uses the information in the `CloudLBConfig`
+ConfigMap to generate the CoreFile to configure the CoreDNS instance in the pod.
```

Comment:
What resource does the MCO typically look at to populate these files? It seems like we should be using an API object rather than a configmap.

Reply:
In the on-prem case, the CoreDNS pod also contains a LoadBalancer so there was no need to pass that information to MCO. In the case of the Custom DNS feature, Load Balance configuration is happening within the Installer (for API and API-Int) and Ingress (for *.apps). So, this LB IP addresses need to be provided to MCO so that the CoreDNS pods can be configured with these IP addresses.
We are using the ConfigMap to solve 2 puposes. 
1. as a source of LB IP addresses to configure CoreDNS 
2. as a source of config that the customer can use to configure their own DNS when cluster creation has completed.